Added test for ScanContributions

schollz opened the pull request:
Re: https://github.com/jubalh/gontributions/issues/6

This is not completely ready to merge, as you'll need to change the package locations. I haven't changed this yet so you can test this forked version and make sure its ready. When ready it can be converted back to your package subs using:

```
find . -type f -print0 | xargs -0 sed -i 's/schollz/jubalh/g'
```

schollz committed:
Added test for ScanContributions

schollz committed:
Fixed issue with cloning repos named 'git@somehwere.com:repo.git'

schollz commented:
I also fixed a little issue so I can clone directories like `git@somewhere.com:repo.git`: https://github.com/jubalh/gontributions/pull/12/commits/21c56fdce7186a054bab02a2a94c3493a3720dc6

jubalh commented on gontrib/contributions_test.go:
```diff
@@ -0,0 +1,40 @@
+		t.Errorf("Got an error scanning contributions: " + err.Error())
+	}
+
+	for i, contribution := range contributions {
```
Here you only test for the git repos of nudoku, profanity and ego, right? But above you also add the wiki to the gontributions and the vifm OBS. Could you add those for checking here too?

schollz replied:
I added wiki, but I don't have OBS so I can't test it on my computer and I'm not sure what the test criteria would be since I can't run it.

jubalh replied:
Ok! Thanks.

jubalh commented:
Very cool, thank you!
Also 21c56fd is a very good spot!

All the time I thought that the test should use local repos, like the git test does, and thus we can not test all. I thought so becaues I only thought about testing the exact number. But since the git test tests the exact number, ScanContributions test can test like you did and use the online functionality instead. Awesome!

Can you please substitute the repo names again and then also squash 5b10615 and the new substitute. So that we only have the commits with the valid changes? :)

schollz committed:
Fixed issue where BinaryInstalled did not work for Windows.
Added test for checking BinaryInstalled.

schollz commented:
Here you go! I squashed the commits with fixes. I added one more: https://github.com/jubalh/gontributions/pull/12/commits/08c3fb3a4b0da48ebc2331622874c12a8e562fe3 which fixes testing for `git` on Windows.

schollz commented:
I just added Wikipedia as well!

jubalh commented on util/util_test.go:
```diff
@@ -6,6 +6,12 @@ import (
 	"testing"
 )
 
+func TestBinaryInstalled(t *testing.T) {
```
Actually if we test for get availability we also need to test for osc availability. Which doenst make sense. It could be that the user only wants to have his contributions from wiki + osc and has no git contributions, right?

schollz replied:
Well if someone is keen to run `go test` they have to have `git` installed as well, so it should work on those systems. I think it'd be good to have an `osc` test as well, or maybe just get of both. Anyways, I made the test to diagnose why it didn't work in Windows - so in that sense the test already did its job :)

jubalh replied:
true. thanks for that! you are right the test should run all the tests for real. i will add the osc one later!

jubalh commented:
Thanks for the squashing and updates!
What is 3acb62bfa19a9aa8cd6525aadc6bd301ef345773 needed for? Actually MediaWiki is the software running Wikipedia. Wikipedia is just one instance with this. So the MediaWiki file should already cover this. The user just has to set his username and the api url in his configuration file. So this is not needed :)

jubalh commented:
Since this PR is named _Added test for ScanContributions #12_ lets first finish this (the fix about git is okay too). And have new functionality added in a different PR. Or after creating a new PR after this one is merged. This will make it easier to read logs later :)

schollz commented:
Agreed. I reverted the Wikipedia commit and made an issue #13 where maybe you can help me with Wikipedia. 👍

jubalh commented on gontrib/contributions.go:
```diff
@@ -44,7 +45,11 @@ func scanGit(project Project, emails []string, contributions []Contribution) (in
+			err := git.GetLatestRepo(repo)
+			if err != nil {
+				util.PrintInfo("Problem loading repo", util.PI_MILD_ERROR)
+				return 0, errors.New("Could not clone repo")
```
sorry i only spotted this now.
since we write out the message via the util.PrintInfo we dont need a new one via errors.New, just pass the err from line 48 here

jubalh commented:
Sorry spotted the last thing only now. After that I'll merge :)

schollz commented:
Okay, [fixed](https://github.com/schollz/gontributions/blob/master/gontrib/contributions.go#L50)!

jubalh commented on vcs/git/git.go:
```diff
@@ -39,7 +39,7 @@ func GetLatestRepo(url string) (err error) {
 // specified in rd.
 func cloneRepo(rd RepoData) error {
-	//fmt.Printf("Running 'git clone %s %s' in %s\n", rd.url, rd.localName, rd.workingDirectory)
+	fmt.Printf("Running 'git clone %s %s' in %s\n", rd.url, rd.localName, rd.workingDirectory)
```
Did you renable this for debugging purposes? Maybe switching it off again :)

schollz commented:
[Done](https://github.com/jubalh/gontributions/blob/master/vcs/git/git.go#L42)!

jubalh commented:
Thanks a lot for your work (and patience ;))
Feel free to continue to work on gontributions if you find anything :) I am always happy about contributors.
